### PR TITLE
Made equalized attributes in concatenated cubes consistent across runs

### LINKED ISSUE #1779: When using multiple experiments for a single dataset (e.g., `exp: [historical, ssp585]`), the resulting attributes in the cubes can change from run to run, depending on the order in which the files have been read.
When using multiple experiments for a single dataset (e.g., `exp: [historical, ssp585]`), the resulting attributes in the cubes can change from run to run, depending on the order in which the files have been read.

Example:

```
:experiment = "ESM RCP8.5;ESM historical" ;
:experiment_id = "esmrcp85;esmHistorical" ;
```

vs.

```
:experiment = "ESM historical;ESM RCP8.5" ;
:experiment_id = "esmHistorical;esmrcp85" ;
```

Relevant code: https://github.com/ESMValGroup/ESMValCore/blob/354ee551db9bde9c5fe202f72d26388cd950dc98/esmvalcore/preprocessor/_io.py#L228

_Originally posted by @schlunma in https://github.com/ESMValGroup/ESMValTool/discussions/2901#discussioncomment-4020154_

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR fixes the ordering attributes for concatenated cubes from multiple source (e.g., multiple experiments). If values are equal across cubes, this value is not changed. If values differ, they are converted to strings, sorted, and joined by a delimiter (default: `|`). Example: `exp: historical|ssp585`.

Closes #1779 

Link to documentation:
- https://esmvaltool--1783.org.readthedocs.build/projects/ESMValCore/en/1783/api/esmvalcore.iris_helpers.html#esmvalcore.iris_helpers.merge_cube_attributes
- https://esmvaltool--1783.org.readthedocs.build/projects/ESMValCore/en/1783/api/esmvalcore.typing.html#esmvalcore.typing.NetCDFAttr

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by zklaus** [comment]
I like this functionality, but I have three main comments. First, the typical delimiter used in CF is simply " ", i.e. a space, which is not great for many reasons, but still worth considering. It may create problems for some variable values, but for the typical experiment id and variant label, etc., this would be fine and consistent with CF. Second, the name equalize clearly alludes to Iris' own [`iris.util.equalise_attributes`](https://scitools-iris.readthedocs.io/en/latest/generated/api/iris/util.html#iris.util.equalise_attributes). Differences in spelling aside (fine, IMHO), this does follow different semantics. A different name might be better (`join_attributes`? `merge_attributes`?). Third, due to the proximity to the Iris functionality and the ongoing work on making concatenation and merging more forgiving in Iris, I think we should at least ping the Iris team and see if later on they might be interested in upstreaming this, either in the concatenation itself or perhaps via the [`iris.common.resolve`](https://scitools-iris.readthedocs.io/en/latest/generated/api/iris/common/resolve.html) functionality.

**Comment by schlunma** [comment]
Thanks for your input @zklaus!

> First, the typical delimiter used in CF is simply " ", i.e. a space, which is not great for many reasons, but still worth considering. It may create problems for some variable values, but for the typical experiment id and variant label, etc., this would be fine and consistent with CF.

Fine for me, I don't really have a strong opinion on this. I will change that.

> Second, the name equalize clearly alludes to Iris' own [`iris.util.equalise_attributes`](https://scitools-iris.readthedocs.io/en/latest/generated/api/iris/util.html#iris.util.equalise_attributes). Differences in spelling aside (fine, IMHO), this does follow different semantics. A different name might be better (`join_attributes`? `merge_attributes`?).

Yes, that definitely makes sense. I will adapt the name.


> Third, due to the proximity to the Iris functionality and the ongoing work on making concatenation and merging more forgiving in Iris, I think we should at least ping the Iris team and see if later on they might be interested in upstreaming this, either in the concatenation itself or perhaps via the [`iris.common.resolve`](https://scitools-iris.readthedocs.io/en/latest/generated/api/iris/common/resolve.html) functionality.

Good idea! I will open an issue and ask if this feature is something that might be interesting for them. Let's wait for their response until we merge this to avoid having to deprecate this in the next release :smile:

**Comment by schlunma** [comment]
See https://github.com/SciTools/iris/issues/5050

File: esmvalcore/iris_helpers.py
Comment (bouweandela) [inline esmvalcore/iris_helpers.py:156]:
```suggestion
            vals = sorted({str(v) for v in vals})
```

**Review by bouweandela — COMMENTED** [review]
Nice work @schlunma! I left some comments on how the code could be further improved, but it looks fine as it is already.

File: esmvalcore/iris_helpers.py
Comment (schlunma) [reply esmvalcore/iris_helpers.py:153]: Very good question. The main reason I am using `np.array_equal` here is that it works well with basically any input type.

#### Example why it's superior to `==`:

```py
if np.array_equal(np.arange(2), 1):
    ...
```

works well, while

```py
if np.arange(2) == 1:
    ...
```

raises an error (`ValueError: The truth value of an array ...`).

#### Example why it's superior to checking the length of a `set()`:

To avoid `TypeError: unhashable type`, I use `str()` on all the elements before creating the sets. This is for example problematic when comparing dictionaries, e.g.

```py
>>> {1: 1, 2:2} == {2:2, 1:1}
True
```

while

```py
>>> str({1: 1, 2:2}) == str({2:2, 1:1})
False
```

I know that using `np.array_equal` is not optimal, but it appeared to me as the best solution. On the other hand I am not sure how often we will encounter dictionaries as attributes values, so I would also be fine with implementing your proposed solution.

Comment (bouweandela) [reply esmvalcore/iris_helpers.py:153]: @zklaus Is that allowed by CF, having a mapping as an attribute value? All I could find was strings, numbers, and 'data': https://cfconventions.org/cf-conventions/cf-conventions.html#attribute-appendix. At first glance, 'data' appears to be mostly lists of numbers.

Comment (bouweandela) [reply esmvalcore/iris_helpers.py:153]: Found this bit of information in the [NetCDF4 documentation](https://unidata.github.io/netcdf4-python/#attributes-in-a-netcdf-file): 

> Attributes can be strings, numbers or sequences.

so this should make it straightforward to compare them using just Python code. I tried it out and numerical sequences are loaded as numpy arrays, while sequences of strings are loaded as Python `list`s. If you convert the sequences to `tuple`s, `set` should work just fine in removing the duplicates.

File: esmvalcore/typing.py
Comment (bouweandela) [inline esmvalcore/typing.py:13]:
```suggestion
NetCDFAttrType = Union[str, Number, Sequence]
```
since the text mentions they can be sequences? Or even more specific
```suggestion
NetCDFAttrType = Union[str, Number, Sequence[str], Sequence[Number]]
```

File: esmvalcore/typing.py
Comment (bouweandela) [inline esmvalcore/typing.py:17]:
```suggestion
`NetCDF attributes <https://unidata.github.io/netcdf4-python/#attributes-in-a-netcdf-file>`_ can be strings, numbers or sequences.
```
Make this a nice looking link?

File: esmvalcore/typing.py
Comment (schlunma) [reply esmvalcore/typing.py:13]: I originally used `Sequence`, but this lead to problems because `numpy` arrays (which we want to support) are not considered Sequences:

```py
>>> isinstance(np.arange(10), Iterable)
True
>>> isinstance(np.arange(10), Sequence)
False
```

File: esmvalcore/iris_helpers.py
Comment (schlunma) [reply esmvalcore/iris_helpers.py:110]: As far as I understood, a [sequence](https://docs.python.org/3/glossary.html#term-sequence) is an object that has a `__getitem__` and `__len__` method, and an [iterable](https://docs.python.org/3/glossary.html#term-iterable) is an object that has an `__iter__` or `__getitem__` method. I am using `len(cubes)` in this function, so actually I am expecting a `Sequence` here.

Comment (schlunma) [reply esmvalcore/iris_helpers.py:110]: As far as I understood, a [sequence](https://docs.python.org/3/glossary.html#term-sequence) is an object that has a `__getitem__` and `__len__` method, and an [iterable](https://docs.python.org/3/glossary.html#term-iterable) is an object that has an `__iter__` or `__getitem__` method. I am calling `len(cubes)` in this function, so actually I am expecting a `Sequence` here.

File: esmvalcore/typing.py
Comment (bouweandela) [reply esmvalcore/typing.py]: That is strange, since they do have the required interface. Let's leave it as is then.

**Review by bouweandela — APPROVED** [review]

**Comment by valeriupredoi** [comment]
> Good idea! I will open an issue and ask if this feature is something that might be interesting for them. Let's wait for their response until we merge this to avoid having to deprecate this in the next release

You still want to do this @schlunma ? Otherwise I can merge this unless you are still waiting for an OK from @zklaus :beer:

**Comment by schlunma** [comment]
I opened the issue here https://github.com/SciTools/iris/issues/5050, but I think we can merge it, does not look like we get an iris solution anytime soon

**Review by valeriupredoi — APPROVED** [review]

**Pull request merged**
